cips/cip-040.md

```diff
@@ -0,0 +1,142 @@
+
+The proposed extension addresses this gap by allowing validators to sign arbitrary message hashes while maintaining the security properties of their existing key management infrastructure.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Protocol Buffer Definitions
+
+The following message types SHALL be added to the privval.Message interface:
+
+```protobuf
+message SignP2PMessageRequest {
+  bytes  hash      = 1;
+  string chain_id  = 2;
+  string unique_id = 3;
+}
+
+message SignedP2PMessageResponse {
+  bytes             signature = 1;
+  RemoteSignerError error     = 2;
+}
+```
+
+### Interface Extension
+
+The privval.Message interface SHALL be extended to include the new message types:
+
+```protobuf
+message Message {
+  oneof sum {
+    PubKeyRequest            pub_key_request             = 1;
+    PubKeyResponse           pub_key_response            = 2;
+    SignVoteRequest          sign_vote_request           = 3;
+    SignedVoteResponse       signed_vote_response        = 4;
+    SignProposalRequest      sign_proposal_request       = 5;
+    SignedProposalResponse   signed_proposal_response    = 6;
+    PingRequest              ping_request                = 7;
+    PingResponse             ping_response               = 8;
+    SignP2PMessageRequest    sign_p2p_message_request    = 9;
+    SignedP2PMessageResponse signed_p2p_message_response = 10;
+  }
+}
+```
+
+### Field Specifications
+
+- `hash`: The SHA-256 hash of the message to be signed. This MUST be exactly 32 bytes.
+- `chain_id`: The chain identifier to prevent cross-chain signature reuse.
+- `unique_id`: An optional identifier for the specific protocol or message type being signed.
+- `signature`: The resulting signature bytes from the signing operation.
+- `error`: Error information if the signing operation fails.
+
+### Implementation Requirements
+
+1. KMS implementations MUST support the new message types for full compatibility.
+2. The signing operation MUST use the same cryptographic key as consensus message signing.
+3. Implementations MUST validate that the hash field is exactly 32 bytes.
+4. The chain_id field MUST match the configured chain identifier.
+5. Double-signing protection is NOT REQUIRED for P2P message signing operations.
+
+## Rationale
+
+### Design Decisions
+
+**Hash-based Signing**: The interface accepts pre-hashed messages rather than raw message content to:
+
+- Maintain consistent message size regardless of original content length
+- Ensure cryptographic properties of the signed data
+- Simplify KMS implementations by avoiding message content validation
```

Comment:
Does each validator get the raw message in addition to the hash of the message before signing? 

If not, is the validator blindly signing over a hash without access to the raw message?